### ISSUE TITLE
lsd 0.21.0 no longer respects terminal color scheme

### ISSUE THREAD

**Issue by OldWorldOrdr** [Peltoche/lsd#609]
- os: Windows 10 (Arch Linux in WSL)
- `lsd --version`:  0.21.0
- `echo $TERM`: xterm-256color
- `echo $LS_COLORS`: rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.webp=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:

## Expected behavior
If applicable, add the output of the classic ls command (`\ls -la`) in order to show the buggy file/directory.

lsd should respect my terminal color scheme, as it does in 0.20 and 0.19

## Actual behavior
If the application panics run the command with the trace (`RUST_BACKTRACE=1 lsd ...`).
In case of graphical errors, add a screenshot if possible.

lsd does not respect my terminal color scheme

![Screenshot (9)](https://user-images.githubusercontent.com/57970114/149818121-4e966d27-8334-4978-8f93-1e8a4c0a22b9.png)

**Comment by zwpaper** [comment]
we change the color lib to a new one for the theme feature, maybe there is some color difference between the 2 libs.

please unset the `LS_COLORS` and let's make sure it is not working or there is a color difference.

**Comment by OldWorldOrdr** [comment]
The issue is still present when I unset LS_COLORS

**Comment by zwpaper** [comment]
nothing changed for both `0.20.1` and `0.21.0`?

it seems like the difference between the 2 default colors.

how are the colors present if the `LS_COLOR` you mentioned is set?

**Comment by OldWorldOrdr** [comment]
what I originally thought was the issue (the new version didnt obey my terminal color scheme) was incorrect, what seems to be happening is that 0.21.0 is using the darker versions of the terminal colors rather than the lighter ones that the old version uses

![Screenshot (13)](https://user-images.githubusercontent.com/57970114/149868736-a3ab0b54-e3ab-4a03-961c-c6a0b4f8c65a.png)

**Comment by zwpaper** [comment]
my point is that even you unset the `LS_COLORS`, the old version still shows the color you expected, so maybe this is not related to `LS_COLORS`, it is caused by the color lib changes.

some minor differences happened when we are changing the lib, maybe this is some cases like that.

maybe we need to figure out why the `LS_COLORS` not working, but it seems really complicated...

**Comment by meain** [comment]
@zwpaper One possibility that I can think of is https://github.com/Peltoche/lsd/blob/1a63ccced0203f2131ad26682da849b16aa66a5d/src/color.rs#L246. But from what I understand this is correct as per https://github.com/crossterm-rs/crossterm/blob/db956267f80744b1d8ab08dcd7d92ee21b7a60f2/src/style/types/color.rs#L126 .

I am NOT able to reproduce the difference between 19 and 21 though.

**Comment by zwpaper** [comment]
yes, this also recalls me of this red difference when implementing the theme feature.

I was also confused about this issue as the `LS_COLORS` env seems not working, I tried the env but nothing changed in my terminal, but I am not good at LS_COLORS

**Comment by meain** [comment]
`LS_COLORS` works as expected. The reason why you did not see a change is probably because this is close to the default.

**Comment by OldWorldOrdr** [comment]
classic `ls` has the colors that 0.20.1 uses

**Comment by meain** [comment]
I have a feeling that this may be linked to https://github.com/Peltoche/lsd/issues/248

**Comment by meain** [comment]
I think I was able to track down what is going on. 

In `lscolors`, 34 is mapped to `lscolors::Colors::Blue` (https://github.com/sharkdp/lscolors/blob/c6e191b91f637058ba932a4f7b065f2e19d8a1ba/src/style.rs#L245) which was initially a `ansi_term` only lib which also mapped `ansi_term::Color::Blue` to 34(https://github.com/ogham/rust-ansi-term/blob/ff7eba98d55ad609c7fcc8c7bb0859b37c7545cc/src/style.rs#L273).

The crossterm integration came later and for crossterm `crossterm::Color::DarkBlue` , which we map to(https://github.com/Peltoche/lsd/blob/1a63ccced0203f2131ad26682da849b16aa66a5d/src/color.rs#L249) is using 4(https://github.com/crossterm-rs/crossterm/blob/db956267f80744b1d8ab08dcd7d92ee21b7a60f2/src/style/types/color.rs#L130) which matches with what neofetch shows. Even `lscolors` maps to `crossterm::Color::Blue` (https://github.com/sharkdp/lscolors/blob/bc95124f23d2f1c14616e6a97f5ae9e6ab786697/src/style.rs#L61) which is still not 34.

Let me do some more research, but if this is the case we will have to solve this upstream in `lscolors` as well.

**Comment by meain** [comment]
Sorry for the delay, but I did some more digging on this. The 34 comes from `ansi-term` using just 8 colors and it is equivalent to 4 in `crossterm` which expects 256 color support.

ref: https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797

---

With that said, I have a feeling the the latest `lsd` is already doing the right thing. @OldWorldOrdr What does gnu ls give you for color: `\ls --color=always`?
![2022-02-08-20-34-438402](https://user-images.githubusercontent.com/14259816/153014524-e3962b4d-fbbc-4c95-a705-dd252886cbac.png)

**Comment by meain** [comment]
@zwpaper Could you set `export LS_COLORS="di=01;31"` and see if `/tmp/one/lsd-0.20.1-x86_64-unknown-linux-gnu/lsd && \ls --color=always && lsd` gives you the same colors. It won't be the same shade of red as me, but at least they should be same between them.

*What I don't get is how the old one has different colors in the issue screenshot*

![2022-02-08-20-42-413944](https://user-images.githubusercontent.com/14259816/153016226-8baa56cc-6038-4746-81cf-aecd7436bd50.png)

**Comment by meain** [comment]
@OldWorldOrdr Could you try building `lsd` after applying this patch and see if it uses the old colors.
```diff
diff --git a/src/color.rs b/src/color.rs
index b5636bf..fedad32 100644
--- a/src/color.rs
+++ b/src/color.rs
@@ -243,12 +243,12 @@ fn to_content_style(ls: &lscolors::Style) -> ContentStyle {
         },
         lscolors::style::Color::Fixed(n) => Color::AnsiValue(*n),
         lscolors::style::Color::Black => Color::Black,
-        lscolors::style::Color::Red => Color::DarkRed,
-        lscolors::style::Color::Green => Color::DarkGreen,
-        lscolors::style::Color::Yellow => Color::DarkYellow,
-        lscolors::style::Color::Blue => Color::DarkBlue,
-        lscolors::style::Color::Magenta => Color::DarkMagenta,
-        lscolors::style::Color::Cyan => Color::DarkCyan,
+        lscolors::style::Color::Red => Color::Red,
+        lscolors::style::Color::Green => Color::Green,
+        lscolors::style::Color::Yellow => Color::Yellow,
+        lscolors::style::Color::Blue => Color::Blue,
+        lscolors::style::Color::Magenta => Color::Magenta,
+        lscolors::style::Color::Cyan => Color::Cyan,
         lscolors::style::Color::White => Color::White,
     };
     let mut style = ContentStyle {
```

If you are not scared of running a random binary that a stranger on the internet gave you, [here](https://d.meain.io/952a990f-lsd) (`md5: 0b1e358c79fc4b159bf3161de985b784`) is a built version.

**Comment by OldWorldOrdr** [comment]
> @OldWorldOrdr Could you try building `lsd` after applying this patch and see if it uses the old colors.
> 
> ```diff
> diff --git a/src/color.rs b/src/color.rs
> index b5636bf..fedad32 100644
> --- a/src/color.rs
> +++ b/src/color.rs
> @@ -243,12 +243,12 @@ fn to_content_style(ls: &lscolors::Style) -> ContentStyle {
>          },
>          lscolors::style::Color::Fixed(n) => Color::AnsiValue(*n),
>          lscolors::style::Color::Black => Color::Black,
> -        lscolors::style::Color::Red => Color::DarkRed,
> -        lscolors::style::Color::Green => Color::DarkGreen,
> -        lscolors::style::Color::Yellow => Color::DarkYellow,
> -        lscolors::style::Color::Blue => Color::DarkBlue,
> -        lscolors::style::Color::Magenta => Color::DarkMagenta,
> -        lscolors::style::Color::Cyan => Color::DarkCyan,
> +        lscolors::style::Color::Red => Color::Red,
> +        lscolors::style::Color::Green => Color::Green,
> +        lscolors::style::Color::Yellow => Color::Yellow,
> +        lscolors::style::Color::Blue => Color::Blue,
> +        lscolors::style::Color::Magenta => Color::Magenta,
> +        lscolors::style::Color::Cyan => Color::Cyan,
>          lscolors::style::Color::White => Color::White,
>      };
>      let mut style = ContentStyle {
> ```
> 
> If you are not scared of running a random binary that a stranger on the internet gave you, [here](https://d.meain.io/952a990f-lsd) (`md5: 0b1e358c79fc4b159bf3161de985b784`) is a built version.

this worked perfectly.

**Comment by zwpaper** [comment]
it is all the same color in my case 🤣, both dark and light

![image](https://user-images.githubusercontent.com/3764335/153110309-9802d669-21bd-43b9-a4d9-c16b56393f53.png)

![image](https://user-images.githubusercontent.com/3764335/153110368-d549109b-74fa-482c-8ad9-76146e72674d.png)

**Comment by meain** [comment]
@zwpaper Yeah, same for me. In fact, if I switch it using the diff I mentioned it gives me different colors compared to gnu ls. What I have no clue now is how the old has different colors for OP.

@OldWorldOrdr If you pipe `lsd` to `less` (or redirect output to a file) with color always, you will get the actual ansi escape sequences. Could you paste that here. `lsd --color=always -d /tmp`. It will look something like `[38;5;4m[1m/tmp[0m`.

**Comment by OldWorldOrdr** [comment]
@meain 

0.20.1:
```file
[1;34mfolder[0m
[1;32mscript.sh[0m
```

0.21.0:
```
file
[38;5;4m[1mfolder[0m
[38;5;2m[1mscript.sh[0m
```

0.21.0 with patch:
```
file
[38;5;12m[1mfolder[0m
[38;5;10m[1mscript.sh[0m
```

**Comment by zwpaper** [comment]
hi @OldWorldOrdr, could you also paste the GNU ls piped contents with `\ls --color=always`

**Comment by OldWorldOrdr** [comment]
@zwpaper 

```
file
[0m[01;34mfolder[0m
[01;32mscript.sh[0m
```

**Comment by meain** [comment]
@OldWorldOrdr I'm not sure how you configure Window terminal, but could you paste the color configuration.

Ideally from what I understand both `1;34` and `38;5;4` `1` should give you the same color. 
`echo '\x1b[1;34m16 color blue\x1b[0m | \x1b[38;5;4m\x1b[1m256 color dark \x1b[0m | \x1b[38;5;12m\x1b[1m256 color light\x1b[0m'`

![2022-02-09-19-12-023944](https://user-images.githubusercontent.com/14259816/153213080-3ddd6810-2437-441b-ba3a-115356e9562c.png)
.

**Comment by OldWorldOrdr** [comment]
@meain 
![Screenshot (28)](https://user-images.githubusercontent.com/57970114/153273308-0dae8458-7d61-4afd-8332-8c4dcd0eea58.png)

**Comment by meain** [comment]
It just hit me. A lot of terminal emulators enable "Use bright text for bold colors" [option](https://github.com/microsoft/terminal/issues/3613). With this enabled, if something is bold, it automatically makes it bright. This I believe does not work on 256 color pallet. So, in your case, the old version of lsd was trying to show directories in bold, but with dark colors (because of `di=1;34` .. 1 = bold 34 = blue), but Windows terminal automatically makes it bright and that is how to end up with the bright variants of the colors in the old version of `lsd`. Try disabling this option if this is available and see if the colors in both versions match.

**Comment by meain** [comment]
@beedaddy See my previous message [here](https://github.com/Peltoche/lsd/issues/609#issuecomment-1034464182).

You said you are using gnome. Just from that, if you are using gnome-terminal then this checkbox at the bottom is the one that I am talking about in my previous message. See if disabling this will make both of the colors the same.
![image](https://user-images.githubusercontent.com/14259816/153581218-40c1f97c-8af2-4de0-89a9-df5a89962cd8.png)

**Comment by beedaddy** [comment]
Thanks for reminding me about this option. Disabling this option results in the same color: the darker blue.
![Bildschirmfoto von 2022-02-11 12-12-20](https://user-images.githubusercontent.com/3264939/153582049-0cc78c60-c104-4564-aaf3-63d73ac1c152.png)
So the problem is that lsd irgnores the (enabled) option?

**Comment by meain** [comment]
Thanks @beedaddy for verifying the issue. The issue here is that older version of lsd relied on 8/16 colors which would change colors based on if this option is enabled or not. The newer version of `lsd` uses 256 color pallete which does not get affected by this setting.

IMO the "right" color is the dark variant. The ideal fix here is to not do a translation from 8/16 color to 256 automatically, but from what I understand `crossterm` does not currently have support for using 8/16 color pallete.

**Comment by OldWorldOrdr** [comment]
its been nearly a year and lsd still uses the wrong colors for me.  Heres an up to date screenshot
![Screenshot (6)](https://user-images.githubusercontent.com/57970114/205046248-153d8da9-2c39-4b37-a8d3-75896f0ee139.png)

**Comment by OldWorldOrdr** [comment]
is there a way to force it to use bright colors instead of dark bold colors?

**Comment by OldWorldOrdr** [comment]
![Screenshot (8)](https://user-images.githubusercontent.com/57970114/205047822-352e89e9-4fcf-4f1b-aa0a-e0e008179c22.png)
refuses to make text bold

**Comment by meain** [comment]
> is there a way to force it to use bright colors instead of dark bold colors?

You will have to use 256 colors `LS_COLORS` to get bright ones. For bright blue, it would be something like `di=38;5;12`. In general, you can use `38;5;<x>` where x is the color. You can list of all the colors using the following script:

``` bash
for code in {0..255}; do
	echo -n "[38;05;${code}m$(printf %03d $code) "
	[ $((${code} % 16)) -eq 15 ] && echo
done
```

**Comment by OldWorldOrdr** [comment]
@meain ur code didn't work so i fixed it for u
working code:
```sh
for code in {0..255}; do
  printf "\e[38;05;${code}m$(printf %03d $code)\n"
  [ $((${code} % 16)) -eq 15 ] && echo
done
```
